src/v/storage/batch_cache.cc

```diff
@@ -9,14 +9,106 @@
 
 #include "batch_cache.h"
 
+#include "model/adl_serde.h"
 #include "utils/to_string.h"
 #include "vassert.h"
 
 #include <fmt/ostream.h>
 
 namespace storage {
 
-batch_cache::entry_ptr
+batch_cache::range::range(batch_cache_index& index)
+  : _index(index) {
+    auto f = new details::io_fragment(
+      ss::temporary_buffer<char>(range_size), details::io_fragment::empty{});
+    _arena.append_take_ownership(f);
+}
+
+batch_cache::range::range(
+  batch_cache_index& index, const model::record_batch& batch)
+  : _index(index) {
+    add(batch);
+}
+
+model::record_batch batch_cache::range::batch(size_t o) {
+    vassert(_valid, "cannot access invalided batch");
+    auto hdr = reflection::from_iobuf<model::record_batch_header>(
+      _arena.share(o, serialized_header_size));
+    auto buffer = _arena.share(
+      o + serialized_header_size,
+      hdr.size_bytes - model::packed_record_batch_header_size);
+
+    return model::record_batch(
+      hdr, std::move(buffer), model::record_batch::tag_ctor_ng{});
+}
+
+model::record_batch_header batch_cache::range::header(size_t o) {
+    vassert(_valid, "cannot access invalided batch");
+    return reflection::from_iobuf<model::record_batch_header>(
+      _arena.share(o, serialized_header_size));
```

Comment:
`iobuf_const_parser` should work here too. iobuf::share is fast, but it still does a lot like building a new iobuf and one or more new fragments.